engine/src/cmd/cloak.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * cloak.h
+ *
+ * Copyright (C) 2001-2023 Daniel Horn, Benjaman Meyer, Roy Falk, Stephen G. Tuggy,
+ * and other Vega Strike contributors.
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef CLOAK_H
+#define CLOAK_H
+
+#include <string>
+
+#include "energetic.h"
+#include "damageable_layer.h"
+
+enum class CloakingStatus {
+    disabled,
+    damaged,
+    ready,
+    cloaking,
+    cloaked,
+    decloaking
+};
+
+class Cloak
+{
+    friend class Unit;
+
+    CloakingStatus status;
+
+    // How much energy cloaking takes per frame
+    double energy;
+
+    // How fast does this starship cloak/decloak
+    double rate;
+
+    // If this unit cloaks like glass or like fading.
+    // Glass is alpha only. Non glass affects rgb as well.
+    bool glass;
+
+    // Current cloak value. 0 is uncloaked. 1 is fully cloaked.
+    double current;
+
+    // The minimum cloaking value...
+    double minimum;
```

Comment:
I like this - a lot. My only question is, can we keep this to integer math? Or do we have to do it within floating point? If we need to keep it as floating point, then fine; if we can move it to integer math that'd be better. Floating point math generally has a higher cost than integer math, but there's also the precision factor that can cause issues so it'd be better for us to use integer as much as possible. If this isn't immediately doable okay. Not going to block on this, just raising the conversation.

Reply:
I'll take a look. I think so.

Reply:
Sort of, but no. I implemented it but the problem is the simulation_atom_var. Unless you make the integers really large (say 10k or more), 
`rate` * simulation_atom_var < 0 
which is cast to 0 and then and cloak doesn't work.

As it is now, it works and is tested. Let's leave it at that.

Reply:
Thanks for checking. Good information to know.